jemalloc-sys/build.rs

```diff
@@ -28,68 +28,138 @@ fn gnu_target(target: &str) -> String {
+    let out_dir = PathBuf::from(env::var_os("OUT_DIR").expect("OUT_DIR was not set"));
+    println!("TARGET={}", target.clone());
+    println!("HOST={}", host.clone());
+    println!("NUM_JOBS={}", num_jobs.clone());
+    println!("OUT_DIR={:?}", out_dir);
+    let build_dir = out_dir.join("build");
+    println!("BUILD_DIR={:?}", build_dir);
+    let src_dir = env::current_dir().expect("failed to get current directory");
+    println!("SRC_DIR={:?}", src_dir);
+
     let unsupported_targets = [
-        "rumprun", "bitrig", "openbsd", "msvc",
-        "emscripten", "fuchsia", "redox", "wasm32",
+        "rumprun",
+        "bitrig",
+        "openbsd",
+        "msvc",
+        "emscripten",
+        "fuchsia",
+        "redox",
+        "wasm32",
     ];
     for i in &unsupported_targets {
         if target.contains(i) {
             panic!("jemalloc does not support target: {}", target);
         }
     }
 
-    let out_dir = PathBuf::from(env::var_os("OUT_DIR").unwrap());
-    let build_dir = out_dir.join("build");
-    let src_dir = env::current_dir().unwrap();
-
     if let Some(jemalloc) = env::var_os("JEMALLOC_OVERRIDE") {
+        println!("jemalloc override set");
         let jemalloc = PathBuf::from(jemalloc);
-        println!("cargo:rustc-link-search=native={}",
-                 jemalloc.parent().unwrap().display());
+        println!(
+            "cargo:rustc-link-search=native={}",
+            jemalloc.parent().unwrap().display()
+        );
         let stem = jemalloc.file_stem().unwrap().to_str().unwrap();
         let name = jemalloc.file_name().unwrap().to_str().unwrap();
-        let kind = if name.ends_with(".a") {"static"} else {"dylib"};
+        let kind = if name.ends_with(".a") {
+            "static"
+        } else {
+            "dylib"
+        };
         println!("cargo:rustc-link-lib={}={}", kind, &stem[3..]);
-        return
+        return;
     }
 
     fs::create_dir_all(&build_dir).unwrap();
+    // Disable -Wextra warnings - jemalloc doesn't compile free of warnings with
+    // it enabled: https://github.com/jemalloc/jemalloc/issues/1196
+    let compiler = cc::Build::new().extra_warnings(false).get_compiler();
+    let cflags = compiler
+        .args()
+        .iter()
+        .map(|s| s.to_str().unwrap())
+        .collect::<Vec<_>>()
+        .join(" ");
+    println!("CC={:?}", compiler.path());
+    println!("CFLAGS={:?}", cflags);
+
+    let jemalloc_src_dir = out_dir.join("jemalloc");
+    println!("JEMALLOC_SRC_DIR={:?}", jemalloc_src_dir);
+
+    if jemalloc_src_dir.exists() {
+        std::fs::remove_dir_all(jemalloc_src_dir.clone()).unwrap();
+    }
 
-    let compiler = cc::Build::new().get_compiler();
-    let cflags = compiler.args().iter().map(|s| s.to_str().unwrap())
-                         .collect::<Vec<_>>().join(" ");
+    // Copy jemalloc submodule to the OUT_DIR
+    assert!(out_dir.exists(), "OUT_DIR does not exist");
+    let mut cmd = Command::new("cp");
```

Comment:
Instead of executing `cp` could this be implemented in Rust? (there's a few small implementations of `cp_r` floating around that can be copied in)

Reply:
So I've replaced this with the `fs_extra` crate, which has a `dir::copy` function.